navigation/infomax.h

```diff
@@ -0,0 +1,187 @@
+class InfoMax
+  : public VisualNavigationBase
+{
+using MatrixType = Matrix<FloatType, Dynamic, Dynamic>;
+using VectorType = Matrix<FloatType, Dynamic, 1>;
+
+public:
+    InfoMax<Rotater, FloatType>(const cv::Size &unwrapRes,
+                                const MatrixType &initialWeights,
+                                FloatType learningRate = 0.0001)
+      : VisualNavigationBase(unwrapRes)
+      , m_LearningRate(learningRate)
+      , m_Weights(initialWeights)
+    {}
+
+    InfoMax<Rotater, FloatType>(const cv::Size &unwrapRes,
+                                FloatType learningRate = 0.0001)
+      : VisualNavigationBase(unwrapRes)
+      , m_LearningRate(learningRate)
+      , m_Weights(getInitialWeights(unwrapRes.width * unwrapRes.height,
+                                    1 + unwrapRes.width * unwrapRes.height))
+    {}
+
+    virtual void train(const cv::Mat &image) override
+    {
+        VectorType u, y;
+        std::tie(u, y) = getUY(image);
+        trainUY(u, y);
+    }
+
+    FloatType decision(const cv::Mat &image) const
+    {
+        const auto decs = m_Weights * getFloatVector(image);
+        return decs.array().abs().sum();
+    }
+
+    template<class... Ts>
+    auto getHeading(Ts &&... args) const
+    {
+        Rotater rotater(getUnwrapResolution(), getMaskImage(), std::forward<Ts>(args)...);
+        std::vector<FloatType> outputs;
+        outputs.reserve(rotater.max());
+        rotater.rotate([this, &outputs] (const cv::Mat &image, auto, auto) {
+            outputs.push_back(decision(image));
+        });
+
+        const auto el = std::min_element(outputs.begin(), outputs.end());
+        size_t bestIndex = std::distance(outputs.begin(), el);
+        if (bestIndex > outputs.size() / 2) {
+            bestIndex -= outputs.size();
+        }
+        const radian_t heading = units::make_unit<turn_t>((double) bestIndex / (double) outputs.size());
+
+        return std::make_tuple(heading, *el, std::move(outputs));
+    }
+
+    const MatrixType &getWeights() const
+    {
+        return m_Weights;
+    }
+
+#ifndef EXPOSE_INFOMAX_INTERNALS
+    private:
+#endif
+    void trainUY(const VectorType &u, const VectorType &y)
+    {
+        // weights = weights + lrate/N * (eye(H)-(y+u)*u') * weights;
+        const auto id = MatrixType::Identity(m_Weights.rows(), m_Weights.rows());
+        const auto sumYU = (y.array() + u.array()).matrix();
+        const FloatType learnRate = m_LearningRate / (FloatType) u.rows();
+        m_Weights.array() += (learnRate * (id - sumYU * u.transpose()) * m_Weights).array();
+    }
+
+    auto getUY(const cv::Mat &image)
+    {
+        assert(image.type() == CV_8UC1);
+
+        const cv::Size &unwrapRes = getUnwrapResolution();
+        assert(image.cols == unwrapRes.width);
+        assert(image.rows == unwrapRes.height);
+
+        // Convert image to vector of floats
+        const auto u = m_Weights * getFloatVector(image);
+        const auto y = tanh(u.array());
```

Comment:
I think Eigen does some cunning stuff to reduce temporary allocations, but I'm guessing there's still going to be some sizable temporary allocations here. It might be good to make ``u``and ``y`` members.

Reply:
Good idea